directory-connector/src/com/untangle/app/directory_connector/GroupManager.java

```diff
@@ -508,9 +508,12 @@ public void run()
             Map<String,Map<String,Set<String>>> domainsGroupsChildrenCache = new ConcurrentHashMap<>(domains.size());
             Map<String,Map<String,Boolean>> domainsUsersCache = new ConcurrentHashMap<>(domains.size());
             Map<String,Boolean> domainUsersMap = new ConcurrentHashMap<>();
-            for( String domain : domains){
+            for (Map.Entry<String, String> entry : domains.entrySet()) {
```

Comment:
Just now I thought about it, if every single entry in `this.adAdapters` list represents one AD server and we anyway have to iterate all of them, why do we first  get a domainMap from `this.adAdapters`, and use that to get an adapter.
Why couldn't we just iterate  `this.adAdapters` here ?

Reply:
adAdpaters is private member of class ActiveDirectoryManagerImpl, We can't directly use it here. The DomainMap will fetch only the adpaters which are enabled at present.

Reply:
As disscussed instead of fetching domain and then adapter one by one, implemented the logic  to get all enabled adapter in one go and then caching the user and group entry.